### ISSUE TITLE
More tax rules for postage

### ISSUE THREAD

**Issue by InformatiqueProg** [thelia/thelia#1664]
We can only choose a tax rule for postage, in some case you need more..

See
http://thelia.net/forum/viewtopic.php?pid=67125#p67125

**Comment by anoziere** [comment]
Mapped the current mechanism while triaging: postage can only be taxed with the single globally-configured rule (`taxrule_id_delivery_module`), read at one choke point, `AbstractDeliveryModuleWithState::buildOrderPostage()`, and stored as one flat figure on `order` (`postage`, `postage_tax`, `postage_tax_rule_title`). Identical on `2.6` and `main`. A cart mixing goods at different VAT rates therefore gets one blended shipping-tax figure, which cannot express the EU 'shipping VAT follows the goods' rule. Worth knowing: `buildOrderPostage()` already accepts an explicit $taxRuleId override, and CustomDelivery uses it from its own config — the capability exists, it just isn't first-class.

Three options, increasing correctness and cost:

**A. Per-line split (the 2024 proposal above).** Split postage pro-rata across order lines, each taxed at its line's own resolved rate, stored per line (mirroring the existing `order_product_tax` pattern, so the machinery is proven). Fiscally correct, but touches every consumer of `order.postage*` (PDFs, loops, BO, exports) and, on `main` only, also the `cart.postage*` columns — plus a historical-order migration question.

**B. Smarter rule choice, same storage.** Make the $taxRuleId override a first-class per-delivery-module BO setting, and/or add a 'highest rate in cart' strategy. No schema change, addresses the literal 'we can only choose one rule' complaint, but the stored figure stays blended.

**C. Display-only breakdown.** Keep schema and amounts, add a computed pro-rata breakdown at render time (invoice, order detail). Cheap, but only answers transparency, not what is actually collected.

Recommendation: scope any real fix to Thelia 3 (schema changes are still cheap in beta) and start with B, which is small and unblocks the common case; A remains the target if fiscal correctness is required. 2.6 is patch-only, so removing the `thelia-2.6` label.

**Comment by anoziere** [comment]
Follow-up on the mapping above: the tax rule, the readers, and what a real breakdown would cost.

## What the rule actually says

Shipping billed by the seller is not a supply of its own. Article 78(b) of Directive 2006/112/EC puts "incidental expenses, such as commission, packing, transport and insurance costs, charged by the supplier to the customer" inside the taxable amount of the supply itself, and the Court's ancillary-supply doctrine (Card Protection Plan, C-349/96) makes an ancillary service share the rate of the principal one. So shipping follows the goods.

For a cart at a single rate that is unambiguous. For a mixed cart the charge has to be split, and the national statements agree on the basis:

- **France** — BOI-TVA-BASE-10-10-10 § 470: where a lump sum covers items taxed differently, "les redevables doivent ventiler les recettes correspondantes suivant chacun des modes d'imposition applicables". A seller who does not ventilate gets the split done by the administration, or the whole taxed at the highest rate.
- **Germany** — shipping is a *Nebenleistung* and is apportioned in the ratio of the net amounts (*Nettoentgelte*) of the goods it carries.
- **United Kingdom** — VAT Notice 700/24 § 2.2: "An additional charge for delivery will follow the liability of the goods." The notice itself does not spell out mixed baskets; the split then falls under the general fair-and-reasonable apportionment rule.

So the default is **pro rata over the untaxed value of the goods, each share taxed at its own rate**.

It is not the only defensible answer, which is why this is a setting rather than a constant:

1. **Ex-works sales.** BOI-TVA-BASE-10-20-10 § 140-150: when the sale is concluded "aux conditions départ", goods travel at the buyer's risk and the transport charge is a distinct service, taxed on its own at the standard rate. That is exactly what Thelia does today — the current single-rule behaviour is correct for this case, not a bug.
2. **Highest rate on the whole charge.** The conservative fallback the French administration applies itself when no ventilation is produced. Never under-collects, and some shops prefer it to avoid the split entirely.
3. **Basis of the split.** Value is the usual basis and the one Germany states; the UK only requires "fair and reasonable", which leaves room for weight-based splits on heavy low-rate goods.

A complete implementation therefore needs a per-shop strategy — *follow the goods pro rata* / *single rule* / *highest rate in cart* — not a hard-coded formula.

## Who reads postage today

A breakdown turns one scalar into a collection, so here is every consumer, on `main`:

**Written here, and nowhere else**

- `core/lib/Thelia/Module/AbstractDeliveryModuleWithState.php:62` — `buildOrderPostage()`, the single choke point
- `core/lib/Thelia/Action/Cart.php:146-148` — cart columns
- `core/lib/Thelia/Action/Order.php:97-99` — order columns, from `OrderEvent`
- `core/lib/Thelia/Domain/Checkout/Service/CheckoutPaymentService.php:51-53` — cart → order carry-over
- `core/lib/Thelia/Domain/Shipping/Service/DeliverySetupService.php:93-94`, `PostageHandler.php:45-47`, `core/lib/Thelia/Domain/Cart/CartFacade.php:123-125`, `core/lib/Thelia/Action/Coupon.php:243-245`, `:258-260` — reset/estimate paths

**Totals**

- `core/lib/Thelia/Model/Order.php:279-280`, `:332-333` — postage and its tax folded into order totals
- `core/lib/Thelia/Model/Order.php:364-366` — untaxed postage derived by subtraction
- `core/lib/Thelia/Model/Cart.php:183`, `:268`
- `core/lib/Thelia/Module/BaseModule.php:477-479` — amount handed to payment modules

**Display**

- `core/lib/Thelia/Core/Template/Loop/Order.php:359-362` — `POSTAGE`, `POSTAGE_TAX`, `POSTAGE_TAX_RULE_TITLE`
- `core/lib/Thelia/Api/Service/DataAccess/AttributeAccessService.php:242`, `:338-340` — Twig/Flexy access
- `thelia-templates/flexy` — `src/UiComponents/Checkout/Summary/Summary.php:48`, `Summary.html.twig:6`
- `thelia-templates/default-twig` — `src/Service/Order/OrderDetailContextBuilder.php:68-70`, `:112-115`; `order/detail.html.twig:228-231`, `:354`
- `thelia-templates/pdf` — `invoice.html.twig:208-210`, `:227`, `:231`, `:247` (the invoice already prints an untaxed postage, a postage tax and a combined tax total)
- `thelia-templates/email` — `order_confirmation.html.twig:100`, `:105` and the three sibling templates

**API**

- `core/lib/Thelia/Api/Resource/Order.php:433-462` (`postage`, `postageTax`, `postageTaxRuleTitle`, groups `admin:read`/`front:read`)
- `core/lib/Thelia/Api/Resource/DeliveryModuleOption.php:185-202`

**Accounting export**

- `core/lib/Thelia/Domain/DataTransfer/Export/Type/OrderExport.php:41`, `:71-72`, `:86`, `:94-96` — raw SQL, exposed as `shipping_cost` / `shipping_tax` / `shipping_tax_rule_title`
- `core/lib/Thelia/Domain/Customer/Service/CustomerPersonalDataExporter.php:161-162`

**A trap worth writing down before anyone builds on this**: `cart.postage` and `order.postage` carry the same name and the opposite meaning. `core/lib/Thelia/Action/Cart.php:146` stores `amount - amountTax` on the cart, so the cart column is tax-exclusive, while `core/lib/Thelia/Domain/Checkout/Service/CheckoutPaymentService.php:51` adds the tax back when the order is created, so the order column is tax-inclusive — which is why `Order::getUntaxedPostage()` (`core/lib/Thelia/Model/Order.php:364`) has to subtract it again. Any ventilation has to pick a side per table rather than assume one convention.

**Delivery modules** — most already choose their own rate through the fourth argument of `buildOrderPostage()`, each with its own duplicated setting: `ColissimoHomeDelivery.php:285`, `ColissimoPickupPoint.php:267`, `ChronopostHomeDelivery.php:388`, `ChronopostPickupPoint.php:382`, `MondialRelayHomeDelivery.php:392`, `MondialRelayPickupPoint.php:372`, `DpdClassic.php:201`, `CustomDelivery.php:287`. `LocalPickup.php:65` does not, and had no way to.

## Shipped now: the per-carrier rule

#3664 makes that fourth argument first-class instead of a convention each module re-invents. `buildOrderPostage()` moves into `Thelia\Module\OrderPostageBuilderTrait` and resolves the rule in three steps: the id passed explicitly, then the module's own `postage_tax_rule_id` entry in `module_config`, then the shop-wide `taxrule_id_delivery_module`. Steps 1 and 3 are unchanged, so the carriers listed above keep their exact current behaviour; the stateless `AbstractDeliveryModule`, which had no postage builder at all, gets the same one.

That answers the literal complaint — one rule for the whole shop — and costs no schema change, which matters while `local/config/schema.xml` is frozen. It does not ventilate anything: the stored figure is still one blended number.

## The breakdown, when schema changes are possible again

The design that keeps every reader above working is to **leave the three flat columns as the authoritative totals** and add a child table detailing how `postage_tax` splits, mirroring `order_product_tax` (`local/config/schema.xml:1495`) so the machinery is already proven:

```xml
<table name="order_postage_tax" namespace="Thelia\Model">
  <column autoIncrement="true" name="id" primaryKey="true" required="true" type="INTEGER" />
  <column name="order_id" required="true" type="INTEGER" />
  <column name="title" required="true" size="255" type="VARCHAR" />
  <column name="description" type="CLOB" />
  <column defaultValue="0.000000" name="untaxed_amount" required="true" scale="6" size="16" type="DECIMAL" />
  <column defaultValue="0.000000" name="amount" required="true" scale="6" size="16" type="DECIMAL" />
  <foreign-key foreignTable="order" name="fk_order_postage_tax_order_id" onDelete="CASCADE" onUpdate="RESTRICT">
    <reference foreign="id" local="order_id" />
  </foreign-key>
  <index name="idx_order_postage_tax_order_id">
    <index-column name="order_id" />
  </index>
  <behavior name="timestampable" />
</table>
```

One row per rate present in the cart. `untaxed_amount` carries the share of the shipping charge attributed to that rate and `amount` the tax on it, so the rate is implicit and no foreign key to `tax_rule` is frozen into the order — the same snapshot discipline `order_product_tax` follows. `sum(untaxed_amount) + sum(amount)` must equal `order.postage`, and `sum(amount)` must equal `order.postage_tax`, with the rounding remainder assigned to the largest share so the totals stay exact to the cent.

Notes on the shape:

- **Cart side stays flat.** `cart.postage*` is an estimate that is recomputed on every change; the breakdown belongs to the order, which is the fiscal document. Adding `cart_postage_tax` would double the write paths for no gain.
- **`order.postage_tax_rule_title`** keeps its meaning when a single rule applies. With a breakdown it is redundant and can hold the dominant rule's title; the detail is in the child rows.
- **The strategy setting** (pro rata / single rule / highest rate) is a `Config` variable and needs no column.

## What it does to existing orders

Nothing, and that is the point of keeping the flat columns authoritative. Orders placed before the change simply have no `order_postage_tax` rows, which reads as "one blended rate, named in `postage_tax_rule_title`" — exactly what they are. No backfill, no recomputation of a paid invoice, no migration script beyond the `CREATE TABLE`. Every reader in the map above keeps returning the same numbers; only the invoice, the order detail and the accounting export gain a detail block, and only when rows exist.

The costs that remain, honestly:

- The pro-rata split needs the cart at the moment postage is computed. `buildOrderPostage()` receives only an amount, a country and a locale, so either the signature grows or the split moves up into `PostageEstimator` / the cart-to-order transformation. That is the real design decision left.
- `Calculator::computeCartTaxFactor()` (`core/lib/Thelia/Domain/Taxation/TaxEngine/Calculator.php:173`) already spreads cart discounts over rates, but as an unweighted mean of per-line factors rather than a value-weighted one. A postage split must not reuse it as is.
- Coupons that remove postage (`core/lib/Thelia/Action/Coupon.php:243`) have to zero the breakdown as well as the totals.

Suggested split: keep #3664 for the per-carrier rule, and open a dedicated issue for the ventilation carrying the table above, so it can ride the next batched `thelia/config` release rather than block on it.

**Comment by anoziere** [comment]
Closing this one: its title is now satisfied, and what is left deserves a ticket that says what it does.

"More tax rules for postage" is delivered. #3664 gave every delivery module its own postage tax rule, resolved as explicit argument, then the module's own setting, then the shop-wide `taxrule_id_delivery_module`. That setting had no back-office screen, which made the feature unreachable for a merchant; #3695 recorded that hole and it is now filled — the tax rules page lists the activated delivery modules and gives each one a selector, the shop-wide rule staying the default.

The rest of this thread was never about more rules, it was about splitting one postage across several of them: a mixed-rate cart still stores a single blended `order.postage_tax`. That work continues in #3701, which carries over the design validated here — the split as a post-processing step on `DeliveryPostageEvent`, the `order_postage_tax` table, the strategy variable defaulting to today's behaviour, and the list of readers to update.

Ten years of context, thanks to everyone who kept it alive.
